File: cpp/myLeetCode/Dynamic_Programming/knapsack_problem/279. Perfect Squares.cpp
// 279. Perfect Squares
// https://leetcode.com/problems/perfect-squares/

#include <iostream>
#include <vector>
#include <limits>

int numSquares(int n) {
    std::vector<int> dp(n + 1, std::numeric_limits<int>::max());
    dp[0] = 0;
    
    for (int i = 1; i <= n; i++) {
        for (int j = 1; j * j <= i; j++) {
            dp[i] = std::min(dp[i], dp[i - j*j] + 1);
        }
    }
    
    return dp[n];
}

int main() {
    std::vector<int> testCases = {12, 13, 27};

    for (int i = 0; i < testCases.size(); i++) {
        int result = numSquares(testCases[i]);

        std::cout << "For the number " << testCases[i]
                  << ", the minimum number of perfect squares needed is " << result << "." << std::endl;
    }

    return 0;
}

File: C++/代码随想录/Dynamic_Programming/knapsack_problem/279. Perfect Squares.py
def numSquares(n):
    dp = [float('inf')] * (n + 1)
    dp[0] = 0

    for i in range(1, n + 1):
        j = 1
        while j * j <= i:
            dp[i] = min(dp[i], dp[i - j * j] + 1)
            j += 1

    return dp[n]

# 测试用例
print(numSquares(12))  # 输出应该是 3，因为 12 = 4 + 4 + 4
print(numSquares(13))  # 输出应该是 2，因为 13 = 4 + 9
